### PR TITLE
Reset host from Gitlab artifact

### PR DESCRIPTION
**Pull request by martipoe**
Adds a task in order to import database and media files from a Gitlab artifact.

### CODE REVIEW

File: recipe/reset_from_gitlab_artifact.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Deployer;
+
+option(
```
Comment (maikschneider) [inline recipe/reset_from_gitlab_artifact.php]: These options are not needed to be declared when accessed via [ConsoleUtility](https://github.com/xima-media/xima-deployer-extended-typo3/blob/main/recipe/db_init.php#L5)->getOption().

File: recipe/reset_from_gitlab_artifact.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace Deployer;
+use SourceBroker\DeployerExtendedDatabase\Utility\ConsoleUtility as DatabaseConsoleUtility;
```
Comment (maikschneider) [inline recipe/reset_from_gitlab_artifact.php]: Why the alias?

File: recipe/reset_from_gitlab_artifact.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            runLocally('cd ' . $activeDir . ' && rm -f artifacts.zip');
+            runLocally('cd ' . $activeDir . ' && {{local/bin/php}} {{bin/typo3cms}} cache:flush');
+            runLocally('cd ' . $activeDir . ' && {{local/bin/php}} {{bin/typo3cms}} cache:warmup');
+        } else {
+            $verbosity = (new DatabaseConsoleUtility())->getVerbosityAsParameter();
+            run('cd {{release_or_current_path}} && {{bin/php}} {{bin/deployer}} reset:from_gitlab_artifact ' . $verbosity . ' --options="token:' . $gitlabApiToken . ',dumpcode:' . $dumpCode . '" {{argument_host}}');
+        };
+    } else {
+        writeln('Gitlab API URL is invalid: ' . $url);
+        exit(1);
```
Comment (maikschneider) [inline recipe/reset_from_gitlab_artifact.php]: Better use `error()`: https://deployer.org/docs/7.x/api#error

File: recipe/reset_from_gitlab_artifact.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+task('reset:from_gitlab_artifact', function () {
+    // set in deploy.php as https://gitlab.example.org/api/v4/projects/<project-id>/jobs/artifacts/<branch>/download?job=<job-of-artifact>
+    $url = get('reset_gitlab_artifact_url');
+    // Gitlab API token for the repository where the artifact is stored. Can be a project token.
+    $gitlabApiToken = (new DatabaseConsoleUtility())->getOption('token', true);
+    // Database dumpcode that was used during the creation of the Gitlab artifact.
+    $dumpCode = (new DatabaseConsoleUtility())->getOption('dumpcode', true);
+
+    if ( filter_var($url, FILTER_VALIDATE_URL) && preg_match('#jobs\/artifacts\/.+\/download$#', parse_url($url, PHP_URL_PATH)) ) {
```
Comment (maikschneider) [inline recipe/reset_from_gitlab_artifact.php]: Please don't use nested conditions. The URL validation can be a standalone condition.,

**Review by maikschneider — CHANGES_REQUESTED** [review]

File: recipe/reset_from_gitlab_artifact.php (hunk 2 above)
Comment (martipoe) [reply recipe/reset_from_gitlab_artifact.php]: There was a conflict when configuring/testing the task for a single project via its deploy.php - SourceBroker actually has two ConsoleUtilities with different capabilities:
```
use SourceBroker\DeployerExtendedTypo3\Utility\ConsoleUtility;
use SourceBroker\DeployerExtendedDatabase\Utility\ConsoleUtility as DatabaseConsoleUtility;
```

Same issue as here: https://github.com/xima-media/xima-deployer-extended-typo3/blob/60360e6e79eb75f7b72fa2e85680099a27de612c/recipe/check_requirements.php#L5-L6

However, in this task only SourceBroker\DeployerExtendedDatabase\Utility\ConsoleUtility is required so it is possible to not use an alias at all.

File: recipe/reset_from_gitlab_artifact.php (hunk 3 above)
Comment (martipoe) [reply recipe/reset_from_gitlab_artifact.php]: As discussed, no output is provided by this function. However, I will resort to GracefulShutdownException as used by Sourcebroker: https://github.com/sourcebroker/deployer-extended-database/blob/53e3d0fe531d6b7d13d8a55934bb1895101842d7/deployer/db/task/db_push.php#L15

File: recipe/reset_from_gitlab_artifact.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    // set in deploy.php as https://gitlab.example.org/api/v4/projects/<project-id>/jobs/artifacts/<branch>/download?job=<job-of-artifact>
+    $url = get('reset_gitlab_artifact_url');
+    // Gitlab API token for the repository where the artifact is stored. Can be a project token.
+    $gitlabApiToken = (new ConsoleUtility())->getOption('token', true);
+    // Database dumpcode that was used during the creation of the Gitlab artifact.
+    $dumpCode = (new ConsoleUtility())->getOption('dumpcode', true);
+
+    if (!filter_var($url, FILTER_VALIDATE_URL) || !preg_match('#jobs\/artifacts\/.+\/download$#', parse_url($url, PHP_URL_PATH))) {
+        throw new GracefulShutdownException('Gitlab API URL is invalid: "' . $url . '"');
+        return;
```
Comment (maikschneider) [inline recipe/reset_from_gitlab_artifact.php]: Please remove me

**Review by maikschneider — CHANGES_REQUESTED** [review]

**Pull request merged**